### PR TITLE
Enable Helmfile autodiscovery by default

### LINKED ISSUE #998: Add Jsonschema support for Autodiscovery spec
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Right now we do not generate Jsonschema for the Autodiscovery specification which has the two following consequences

* Documentation is not generate on www.updatecli.io
* No auto completion available from IDE

A bit of refactoring is needed in the code as I took some shortcut in the initial implementation to land the autodiscovery feature sooner than later.

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS:
- updatecli:
```

### Anything else?

Similarly to other Updatecli resources, I consider refactor the spec to implement the Unmarshaler but due to https://github.com/go-yaml/yaml/issues/822 inline field are not handled correctly

### PR DESCRIPTION
**Pull request by olblak**
Fix #998 

The situation evolved a lot since I started working on the autodiscovery, and this pullrequest is an attempt to refactor a bit

- [x] Enable Helmfile autodiscovery by default
- [x] Remove Action/scm from the autodiscovery spec
- [x] Remove deprecated flag `local-autodiscovery`
- [x] Remove autodiscovery dependency on Updatecli manifest config
- [x] Fix Jsonschema generator for autodiscovery
- [x] Each auto-discovery plugin returns a manifests as a `[][]byte` instead of `config.Spec`, to fix a dependency cycle issue.

## Test

To test this pull request, you can run the following commands:

```shell
cp <to_package_directory>
go test
```

### Tradeoff

The previous implementation had a dependency cycle between the packages "autodiscovery", "config", and each autodiscovery plugins. To fix the dependency, I had to decouple a bit so now we rely a more on the e2e test to identify future regression issue.

### CODE REVIEW

File: pkg/core/engine/main.go [hunk 1]
```diff
@@ -47,9 +50,9 @@ func (e *Engine) Clean() (err error) {
 // GetFiles return an array with every valid files.
 func GetFiles(root string) (files []string) {
 	if root == "" {
-		// If no manifest have been provided then we try to see if the file
-		// updatecli.yaml exist. If it's then we try to see if the directory updatecli.d
-		// if it's still not the case then we return no manifest files.
+		// Updatecli tries to load the file updatecli.yaml if no manifest provided
+		// If updatecli.yaml doesn't exists then Updatecli parses the directory updatecli.d for any manifests.
+		// if they're no manifests in the directory updatecli.d then Updatecli returns no manifest files.
```
Comment (lemeurherve) [inline pkg/core/engine/main.go:55]:
```suggestion
		// if there is no manifests in the directory updatecli.d then Updatecli returns no manifest files.
```

File: pkg/plugins/autodiscovery/dockerfile/dockerfile.go [hunk 2]
```diff
@@ -142,40 +137,46 @@ func (h Dockerfile) discoverDockerfileManifests() ([]config.Spec, error) {
...
+				TagFilter            string
+				VersionFilterKind    string
+				VersionFilterPattern string
+				ScmID                string
+			}{
+				ManifestName:         fmt.Sprintf("Bump Docker Image Tag for %q", imageName),
+				ImageName:            imageName,
+				SourceID:             imageName,
+				TargetID:             imageName,
+				TargetName:           fmt.Sprintf("[%s] Bump Docker Image tag in %q",imageName,relativeFoundDockerfile),
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/dockerfile/dockerfile.go:164]:
```suggestion
				ManifestName:         fmt.Sprintf("Bump Docker image tag for %q", imageName),
				ImageName:            imageName,
				SourceID:             imageName,
				TargetID:             imageName,
				TargetName:           fmt.Sprintf("[%s] Bump Docker image tag in %q",imageName,relativeFoundDockerfile),
```

File: pkg/plugins/autodiscovery/dockerfile/manifestTemplate.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package dockerfile
+
+var (
+	// manifestTemplate is the Go template used to generate
+	// Updatecli manifests
+	manifestTemplate string = `name: '{{ .ManifestName }}'
+sources:
+  {{ .SourceID }}:
+    name: '[{{ .ImageName }}] Get latest Docker Image Tag'
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/dockerfile/manifestTemplate.go:9]:
```suggestion
    name: '[{{ .ImageName }}] Get latest Docker image tag'
```

File: pkg/plugins/autodiscovery/dockerfile/test/main_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    kind: 'dockerimage'
+    spec:
+      image: 'jenkins/jenkins'
+      tagFilter: '^\d*(\.\d*){2}-lts$'
+      versionFilter:
+        kind: 'semver'
+        pattern: '>=2.235.1-lts'
+targets:
+  jenkins/jenkins:
+    name: '[jenkins/jenkins] Bump Docker Image tag in "Dockerfile"'
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/dockerfile/test/main_test.go:64]:
```suggestion
			expectedPipelines: []string{`name: 'Bump Docker Image Tag for "updatecli/updatecli"'
sources:
  updatecli/updatecli:
    name: '[updatecli/updatecli] Get latest Docker Image Tag'
    kind: 'dockerimage'
    spec:
      image: 'updatecli/updatecli'
      tagFilter: '^v\d*(\.\d*){2}$'
      versionFilter:
        kind: 'semver'
        pattern: '>=v0.25.0'
targets:
  updatecli/updatecli:
    name: '[updatecli/updatecli] Bump Docker Image tag in "Dockerfile"'
    kind: 'yaml'
    spec:
      file: 'Dockerfile'
      instruction:
        keyword: 'FROM'
        matcher: 'updatecli/updatecli'
    sourceid: 'updatecli/updatecli'
`},
		},
		{
			name:    "Scenario 2: arg with suffix",
			rootDir: "testdata/jenkins",
			expectedPipelines: []string{`name: 'Bump Docker image tag for "jenkins/jenkins"'
sources:
  jenkins/jenkins:
    name: '[jenkins/jenkins] Get latest Docker image tag'
    kind: 'dockerimage'
    spec:
      image: 'jenkins/jenkins'
      tagFilter: '^\d*(\.\d*){2}-lts$'
      versionFilter:
        kind: 'semver'
        pattern: '>=2.235.1-lts'
targets:
  jenkins/jenkins:
    name: '[jenkins/jenkins] Bump Docker image tag in "Dockerfile"'
```

File: pkg/plugins/autodiscovery/dockerfile/test/main_test.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+				}, "", "")
+
+			require.NoError(t, err)
+
+			var pipelines []string
+			rawPipelines, err := dockerfile.DiscoverManifests()
+			require.NoError(t, err)
+
+			for i := range rawPipelines {
+				// We expected manifest generated by the autodiscovery to use the yaml syntax
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/dockerfile/test/main_test.go:91]:
```suggestion
				// We expect manifest generated by the autodiscovery to use the yaml syntax
```

File: pkg/plugins/autodiscovery/fleet/dependencies.go [hunk 6]
```diff
@@ -87,67 +82,104 @@ func (f Fleet) discoverFleetDependenciesManifests() ([]config.Spec, error) {
...
+		//				Kind: "yaml",
+		//				Spec: helm.Spec{
+		//					File: relativeFoundChartFile,
+		//					Key:  "helm.version",
+		//				},
+		//			},
+		//		},
+		//	},
+		//}
+
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/fleet/dependencies.go:139]:
```suggestion
```

File: pkg/plugins/autodiscovery/fleet/dependencies.go [hunk 7]
```diff
@@ -87,67 +82,104 @@ func (f Fleet) discoverFleetDependenciesManifests() ([]config.Spec, error) {
...
+			File                       string
+			ScmID                      string
+		}{
+			ManifestName:               fmt.Sprintf("Bump Fleet Bundle %q for Helm Chart %q", chartName, data.Helm.Chart),
+			ChartName:                  data.Helm.Chart,
+			ChartRepository:            data.Helm.Repo,
+			ConditionID:                data.Helm.Chart,
+			FleetBundle:                chartName,
+			SourceID:                   data.Helm.Chart,
+			SourceName:                 fmt.Sprintf("Get latest %q Helm Chart Version", data.Helm.Chart),
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/fleet/dependencies.go:168]:
```suggestion
			ManifestName:               fmt.Sprintf("Bump %q Fleet bundle for %q Helm chart", chartName, data.Helm.Chart),
			ChartName:                  data.Helm.Chart,
			ChartRepository:            data.Helm.Repo,
			ConditionID:                data.Helm.Chart,
			FleetBundle:                chartName,
			SourceID:                   data.Helm.Chart,
			SourceName:                 fmt.Sprintf("Get latest %q Helm chart version", data.Helm.Chart),
```

File: pkg/plugins/autodiscovery/fleet/manifestTemplate.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package fleet
+
+const (
+	// manifestTemplate is the Go template used to generate
+	// Docker compose manifests
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/fleet/manifestTemplate.go:5]:
```suggestion
	// manifestTemplate is the Go template used to generate Fleet manifests
```

File: pkg/plugins/autodiscovery/fleet/test/main_test.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+	testdata := []struct {
+		name              string
+		rootDir           string
+		expectedPipelines []string
+	}{
+		{
+			name:    "Scenario 1",
+			rootDir: "testdata",
+			expectedPipelines: []string{`name: 'Bump Fleet Bundle "grafana" for Helm Chart "grafana"'
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/fleet/test/main_test.go:25]:
```suggestion
			expectedPipelines: []string{`name: 'Bump "grafana" Fleet bundle for "grafana" Helm chart'
```
Is a fleet included in a chart or the inverse?

File: pkg/plugins/autodiscovery/fleet/test/main_test.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+				}, "", "")
+
+			require.NoError(t, err)
+
+			var pipelines []string
+			rawPipelines, err := fleet.DiscoverManifests()
+			require.NoError(t, err)
+
+			for i := range rawPipelines {
+				// We expected manifest generated by the autodiscovery to use the yaml syntax
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/fleet/test/main_test.go:80]:
```suggestion
				// We expect manifest generated by the autodiscovery to use the yaml syntax
```

File: pkg/plugins/autodiscovery/helmfile/main.go [hunk 11]
```diff
@@ -30,10 +28,12 @@ type Helmfile struct {
 	spec Spec
 	// rootdir defines the root directory from where looking for Helmfile
 	rootDir string
+	// scmID hold the scmID used by the newly generated manifest
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/helmfile/main.go:31]:
```suggestion
	// scmID holds the scmID used by the newly generated manifest
```

File: pkg/plugins/autodiscovery/maven/main.go [hunk 12]
```diff
@@ -28,10 +26,12 @@ type Maven struct {
 	spec Spec
 	// rootDir defines the root directory from where looking for Helm Chart
 	rootDir string
+	// scmID hold the scmID used by the newly generated manifest
```
Comment (lemeurherve) [inline pkg/plugins/autodiscovery/maven/main.go:29]:
```suggestion
	// scmID holds the scmID used by the newly generated manifest
```

**Review by lemeurherve — COMMENTED** [review]
First review, mostly on the form

**Comment by olblak** [comment]
@hervelemeur Thanks for the review but this PR is not ready yet :D 
I am still cleaning/removing some code then I'll have to review the comments

File: pkg/core/pipeline/autodiscovery/main.go [hunk 13]
```diff
@@ -1,60 +1,50 @@
...
 	// DefaultGenericsSpecs defines the default builder that we want to run
-	DefaultCrawlerSpecs = discoveryConfig.Config{
+	DefaultCrawlerSpecs = Config{
 		Crawlers: map[string]interface{}{
 			"dockercompose": dockercompose.Spec{},
 			"dockerfile":    dockerfile.Spec{},
-			"helm":          helm.Spec{},
-			"maven":         maven.Spec{},
+			//"helm":          helm.Spec{},
+			//"maven":         maven.Spec{},
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/pipeline/autodiscovery/main.go:20]: It's commented while I fix the cycle dependency issue for those plugins

File: pkg/core/pipeline/autodiscovery/main.go [hunk 14]
```diff
@@ -1,60 +1,50 @@
...
 			"rancher/fleet": fleet.Spec{},
 		},
 	}
+	// AutodiscoverySpecs is a map of all Autodiscovery specification
+	AutodiscoverySpecsMapping = map[string]interface{}{
+		"dockercompose": &dockercompose.Spec{},
+		"dockerfile":    &dockerfile.Spec{},
+		//"helm":          &helm.Spec{},
+		//"maven":         &maven.Spec{},
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/pipeline/autodiscovery/main.go:29]: Same here

File: pkg/core/pipeline/autodiscovery/main.go [hunk 15]
```diff
@@ -65,34 +55,21 @@ func New(spec discoveryConfig.Config,
...
 			continue
 		}
 
-		// Retrieve the scm workdir if it exist
-		// As long as the autodiscovery specifies one
-		if _, ok := g.spec.Crawlers[kind]; ok && scmHandler != nil {
-			workDir = scmHandler.GetDirectory()
-		}
-
+		// Commmenting for now while refactoring
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/pipeline/autodiscovery/main.go:65]: I am still refactoring :D

**Comment by lemeurherve** [comment]
Oups, sorry!

**Comment by olblak** [comment]
@hervelemeur thanks for the early review :D

**Comment by olblak** [comment]
@hervelemeur now it should be ready for review :) 
I fixed a bunch of typos in the process

**Comment by olblak** [comment]
Well this refactoring is really painful :D

**Comment by olblak** [comment]
I am ready to merge this pullrequest.
During the refactoring I noticed many (minor) issues which I addressed in this pullrequest but I still have many wording to improve which I'll do later

File: pkg/plugins/autodiscovery/dockercompose/test/testdata/docker-compose.yaml [hunk 16]
```diff
@@ -7,7 +7,7 @@ services:
...
     ports:
       - "8080:8080"
-  jenkins-lts:
-    image: jenkinsci/jenkins:2.150.1-alpine
-    ports:
-      - "8080:8080"
+  #jenkins-lts:
+  #  image: jenkinsci/jenkins:2.150.1-alpine
+  #  ports:
+  #    - "8080:8080"
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/plugins/autodiscovery/dockercompose/test/testdata/docker-compose.yaml]: I reenabled the snippet

**Pull request merged**